enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,782 @@
+
+When the Machine API controller attempts to reconcile an object, it must first check for the presence of the field.
+If it is not present, then the Machine API controller should reconcile the object.
+It is expected that the sync controller will default it to `machine.openshift.io` in this case.
+
+We will implement predicates/event filters for both Machine API and Cluster API controllers.
+
+Machine API controllers will be allowed to reconcile any object that is either:
+
+* Missing the authoritative field in the status (as today)
+* Has the authoritative field set with the `machine.openshift.io` value
+
+Cluster API controllers will be allowed to reconcile any object that either:
+
+* Has a Machine API equivalent AND has the authoritative field set with the `cluster.x-k8s.io` value
+* Does not have a Machine API equivalent
+
+By following the above rules, we should be able to ensure that the authoritative annotation is present in only one location and that the reconciliation of the objects is well defined based on the presence of this annotation.
+
+The APIs will be extended as follows:
+
+```go
+type xxxSpec struct {
+  // authoritativeAPI is the API that is authoritative for this resource.
+  // Valid values are machine.openshift.io and cluster.x-k8s.io.
+  // When set to machine.openshift.io, writes to the spec of the machine.openshift.io copy of this resource will be reflected into the cluster.x-k8s.io copy.
+  // When set to cluster.x-k8s.io, writes to the spec of the cluster.x-k8s.io copy of this resource will be reflected into the machine.openshift.io copy.
+  // Updates to the status will be reflected in both copies of the resource, based on the controller implementing the functionality of the API.
+  // Currently the authoritative API determines which controller will manage the resource, this will change in a future release.
+  // To ensure the change has been accepted, please verify that the `status.authoritativeAPI` field has been updated to the desired value and that the `Synchronized` condition is present and set to `True`.
+  // +kubebuilder:validation:Enum=machine.openshift.io;cluster.x-k8s.io
+  // +kubebuilder:validation:Default:=machine.openshift.io
+  // +default:=machine.openshift.io
+  // +optional
+  AuthoritativeAPI string `json:"authoritativeAPI,omitempty"`
+}
+
+type xxxStatus struct {
+  // authoritativeAPI is the API that is authoritative for this resource.
+  // Valid values are machine.openshift.io and cluster.x-k8s.io.
+  // This value is updated by the sync controller to reflect the authoritative API.
+  // Machine API and Cluster API controllers use this value to determine whether or not to reconcile the resource.
+  // +kubebuilder:validation:Enum=machine.openshift.io;cluster.x-k8s.io
+  // +optional
+  AuthoritativeAPI string `json:"authoritativeAPI,omitempty"`
+
+  // synchronizedGeneration is the generation of the authoritative resource that the non-authoritative resource is synchronised with.
```

Comment:
This is a bit confusing, expand?

Reply:
Expanded the comment, does that help at all with the understanding?